feat(ctb): Clock extensions

clabby opened the pull request:
## Overview

Introduces clock extensions into the `FaultDisputeGame`, which allow for bounded time credit when having to inherit a clock during responding to freeloader claims.

clabby committed:
feat(ctb): Clock extensions

clabby commented:
* **#10199** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/10199?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#10182** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/10182?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `develop`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @clabby and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

clabby committed:
Add configurable clock extension immutable

locks,bindings,etc.

comment fixes

x

refcell commented on packages/contracts-bedrock/test/dispute/FaultDisputeGame.t.sol:
```diff
@@ -67,7 +68,7 @@ contract FaultDisputeGame_Init is DisputeGameFactory_Init {
         assertEq(gameProxy.maxGameDepth(), 2 ** 3);
         assertEq(gameProxy.splitDepth(), 2 ** 2);
-        assertEq(gameProxy.gameDuration().raw(), 7 days);
+        assertEq(gameProxy.maxClockDuration().raw(), 3.5 days);
```
If it's too annoying to regen bindings, in a follow on pr let's add a check here for the clock extension being set to `3 hours`.

clabby replied:
Nah let's do that now, good callout.

clabby committed:
`clockExtension` config checks

clabby commented:
Adding https://github.com/ethereum-optimism/optimism/labels/M-do-not-merge, would like a second reviewer on the Go changes.

Inphi commented on packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol:
```diff
@@ -261,9 +270,22 @@ contract FaultDisputeGame is IFaultDisputeGame, Clone, ISemver {
+        if (nextDuration.raw() > MAX_CLOCK_DURATION.raw() - CLOCK_EXTENSION.raw()) {
+            // If the potential grandchild is an execution trace bisection root, double the clock extension.
+            uint64 extensionPeriod =
+                nextPositionDepth == SPLIT_DEPTH - 1 ? CLOCK_EXTENSION.raw() * 2 : CLOCK_EXTENSION.raw();
```
Shouldn't this be `SPLIT_DEPTH + 1`?

clabby replied:
idts - the clock extension is for the potential grandchild. Doing this @ `SPLIT_DEPTH + 1` would imply the clock extension applies to the 3rd level in the execution trace sub-tree.

Inphi replied:
got it. I originally thought we'd want to clock extend at the time the grandchild moves. But doing so  allows the freeloader to resolve their own subgame before the honest player has a chance to dispute it.

Inphi commented on op-challenger/game/fault/agent_test.go:
```diff
@@ -170,7 +170,7 @@ func setupTestAgent(t *testing.T) (*Agent, *stubClaimLoader, *stubResponder) {
 	claimLoader := &stubClaimLoader{}
 	depth := types.Depth(4)
-	gameDuration := 6 * time.Minute
+	gameDuration := 3 * time.Minute
```
Why is this halved? The consequence is the test challenger has at most 90 seconds to respond which might not be enough time in CI.

clabby replied:
This was halved because a bit down stream from this, `NewAgent` now accepts the max clock duration. This variable name should be updated.

Inphi replied:
The `gameDuration` is also used by the `Agent` to [determine](https://github.com/ethereum-optimism/optimism/blob/6f99544c57ef837afaa96f979a6154a4e8a3604a/op-challenger/game/fault/agent.go#L166) when to resolve subgames. We should tweak that logic to consider extended subgames. Otherwise, the challenger will be too eager when resolving claims - nbd, just annoying to look at in our logs.

For now, it's OK to add a TODO referencing an issue fixing `resolveClaims`. We can fix this later.

ajsutton replied:
I don't think I'd bother adding this complexity to the challenger. Once the challenger thinks the clock expires it does an `eth_call` to see if the claim will successfully resolve and only actually sends a transaction if that passes. So it won't even spam the logs [other than at trace level](https://github.com/ethereum-optimism/optimism/blob/6f99544c57ef837afaa96f979a6154a4e8a3604a/op-challenger/game/fault/agent.go#L182C4-L182C81).  Given this should be a very rare occurrence the extra eth_call seems fine.  We only added the clock check quite recently to reduce the amount of time the challenger spent making those eth_calls when there were lots of claims and a relatively slow RPC.

ajsutton replied:
oh, re-reading the above conversation are we saying that the challenger shouldn't be dividing the game duration by 2 anymore?

ajsutton replied:
I see we did indeed remove the / 2 in the agent here so I think we're all good and can just ignore the possibility of clock extension in the challenger.

Inphi commented on op-dispute-mon/mon/bonds/monitor.go:
```diff
@@ -47,7 +47,7 @@ func (b *Bonds) CheckBonds(games []*types.EnrichedGameData) {
 	// Check if the max duration has been reached for this game
 	duration := uint64(b.clock.Now().Unix()) - game.Timestamp
-	maxDurationReached := duration >= game.Duration
+	maxDurationReached := duration >= game.MaxClockDuration
```
This should be at least twice the max clock duration. @refcell would it be more accurate to check if the root clock has expired rather than use `maxDurationReached`?

clabby replied:
Hmm - good q. There's really no "max duration" anymore, without deriving it from the configured depth and the possible clock extension upper bound. We really want to see when the clocks expire.

ajsutton replied:
The alert we want to setup based on this metric is:
> op_dispute_mon_credits{credit="below_expectation", max_duration="unreached"} a claimant has less credit than they should have been granted. Either they have withdrawn earlier than the DelayedWETH contract should have allowed or they were not allocated all the bonds they should have been. Note: the WETH may still be locked even after max_duration changes to reached but we don't want to over-complicate the monitoring to get this perfect - we know withdrawals should be locked at least as long as the max game duration.

So it's already imperfect and just using `game.MaxClockDuration * 2` would be reasonable. As long as the funds wind up locked _longer_ than this duration the alert still works well because it's just trying to identify someone withdrawing prior to the earliest point it should be possible to withdraw.

Inphi commented on op-dispute-mon/mon/resolutions.go:
```diff
@@ -52,7 +52,7 @@ func (r *ResolutionMonitor) CheckResolutions(games []*types.EnrichedGameData) {
 		complete := game.Status != gameTypes.GameStatusInProgress
 		duration := uint64(r.clock.Now().Unix()) - game.Timestamp
-		maxDurationReached := duration >= game.Duration
+		maxDurationReached := duration >= game.MaxClockDuration
```
If clock extension triggers for this game, then the game would no longer be `complete`.

ajsutton replied:
yeah this one might cause some false alarms because games appear to not have resolved in time due to clock extension. I'm tempted to just leave it as-is though because if we're hitting cases where clock extension is required, something funky is going on and we probably should take a look anyway.
The only way I could see to counter it is to have the contract record any time extensions it grants which seems excessive.  We don't want to check the remaining duration on clocks here because we want to detect if someone finds a way to extend the clocks beyond what should be possible.

Inphi reviewed (commented):
For dispute-mon metrics, we don't want to alert prematurely if the game is clock extended. But we also don't want to wait too long before alerting, even if the clock-extended game is live.

Perhaps it's OK to have false positive alerts by relaxing the former since clock extensions should be rare.

clabby committed:
feat(ctb): Enforce max supported depth in `FaultDisputeGame` constructor

## Overview

Enforces the max supported depth of the `LibPosition` library's
implemented functionality on the `Position` type in the
`FaultDisputeGame`'s constructor

ajsutton commented on op-challenger/game/fault/contracts/faultdisputegame.go:
```diff
@@ -274,11 +274,11 @@ func (f *FaultDisputeGameContract) GetOracle(ctx context.Context) (*PreimageOrac
-	defer f.metrics.StartContractRequest("GetGameDuration")()
-	result, err := f.multiCaller.SingleCall(ctx, rpcblock.Latest, f.contract.Call(methodGameDuration))
+func (f *FaultDisputeGameContract) GetMaxClockDuration(ctx context.Context) (time.Duration, error) {
+	defer f.metrics.StartContractRequest("GetMaxClockDuration")()
```
Note that once we merge this we can't update the challenger until we update contracts. Not sure what changes are pending to deploy but might be worth doing one last release prior to merging to get the latest changes live.

ajsutton reviewed (approved):
LGTM.

clabby committed:
use clock dur * 2 in bond monitor
